File: frontend/lib/domain/research-advanced.ts
/**
 * SP-032 (ADR-00052): research advanced の client-safe pure schema / 型 / 表示 helper。
 *
 * conflict groups + conflict candidates + per-claim computed_freshness + domain trust registry。
 * 本 module は next/headers を import しない (Client Component から安全に import 可能)。
 * server fetch は `@/lib/api/research-advanced` を使う。
 */

import { z } from "zod";

export const ConflictGroupStatusEnum = z.enum(["open", "resolved", "dismissed"]);
export type ConflictGroupStatus = z.infer<typeof ConflictGroupStatusEnum>;

export const TrustTierEnum = z.enum(["low", "medium", "high"]);
export type TrustTier = z.infer<typeof TrustTierEnum>;

export const DomainTrustMatchTypeEnum = z.enum(["exact", "none", "invalid"]);
export type DomainTrustMatchType = z.infer<typeof DomainTrustMatchTypeEnum>;

export const ConflictGroupSchema = z.object({
  id: z.string().uuid(),
  tenant_id: z.number().int(),
  project_id: z.string().uuid(),
  research_task_id: z.string().uuid(),
  title: z.string(),
  status: ConflictGroupStatusEnum,
  resolution_note: z.string().nullable(),
  created_by_actor_id: z.string().uuid(),
  created_at: z.string(),
  updated_at: z.string()
});
export type ConflictGroup = z.infer<typeof ConflictGroupSchema>;

export const ConflictCandidateSchema = z.object({
  claim_id: z.string().uuid(),
  contradicting_count: z.number().int(),
  supporting_count: z.number().int(),
  context_count: z.number().int(),
  conflict_group_id: z.string().uuid().nullable()
});
export type ConflictCandidate = z.infer<typeof ConflictCandidateSchema>;

export const ClaimFreshnessSchema = z.object({
  claim_id: z.string().uuid(),
  computed_freshness: z.number().nullable(),
  newest_evidence_at: z.string().nullable()
});
export type ClaimFreshness = z.infer<typeof ClaimFreshnessSchema>;

export const EvidenceDomainTrustSchema = z.object({
  evidence_source_id: z.string().uuid(),
  domain: z.string().nullable(),
  trust_tier: TrustTierEnum.nullable(),
  match_type: DomainTrustMatchTypeEnum
});
export type EvidenceDomainTrust = z.infer<typeof EvidenceDomainTrustSchema>;

export const ResearchAdvancedSummarySchema = z.object({
  research_task_id: z.string().uuid(),
  conflict_groups: z.array(ConflictGroupSchema),
  conflict_candidates: z.array(ConflictCandidateSchema),
  relation_coverage: z.number(),
  claim_freshness: z.array(ClaimFreshnessSchema),
  evidence_domain_trust: z.array(EvidenceDomainTrustSchema)
});
export type ResearchAdvancedSummary = z.infer<typeof ResearchAdvancedSummarySchema>;

export const DomainTrustSchema = z.object({
  id: z.string().uuid(),
  tenant_id: z.number().int(),
  domain: z.string(),
  trust_tier: TrustTierEnum,
  rationale: z.string().nullable(),
  created_by_actor_id: z.string().uuid(),
  created_at: z.string(),
  updated_at: z.string()
});
export type DomainTrust = z.infer<typeof DomainTrustSchema>;

export const DomainTrustListResponseSchema = z.object({
  items: z.array(DomainTrustSchema)
});
export type DomainTrustListResponse = z.infer<typeof DomainTrustListResponseSchema>;

// --- display helpers (pure) ---

export function conflictStatusLabel(status: ConflictGroupStatus): string {
  switch (status) {
    case "open":
      return "未解決";
    case "resolved":
      return "解決済み";
    case "dismissed":
      return "却下";
  }
}

export function conflictStatusTone(status: ConflictGroupStatus): "warning" | "success" | "muted" {
  switch (status) {
    case "open":
      return "warning";
    case "resolved":
      return "success";
    case "dismissed":
      return "muted";
  }
}

export function trustTierLabel(tier: TrustTier): string {
  switch (tier) {
    case "low":
      return "低";
    case "medium":
      return "中";
    case "high":
      return "高";
  }
}

export function trustTierTone(tier: TrustTier): "danger" | "warning" | "success" {
  switch (tier) {
    case "low":
      return "danger";
    case "medium":
      return "warning";
    case "high":
      return "success";
  }
}

export function matchTypeLabel(match: DomainTrustMatchType): string {
  switch (match) {
    case "exact":
      return "登録済み";
    case "none":
      return "未登録";
    case "invalid":
      return "ドメイン判定不能";
  }
}

/** computed_freshness (0-1) を percentage 表示。null は "—"。 */
export function formatFreshness(value: number | null): string {
  if (value === null) {
    return "—";
  }
  const clamped = Math.max(0, Math.min(1, value));
  return `${Math.round(clamped * 100)}%`;
}

File: backend/app/schemas/research_advanced.py
"""SP-032 (ADR-00052): research advanced の read-only response schema。

conflict candidates (deterministic 検出) + per-claim computed_freshness (advisory) +
evidence source ごとの domain trust 適用結果 (match_type 付き) を集約する read model。
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from backend.app.db.models.domain_trust import TrustTier
from backend.app.schemas.conflict_group import ConflictGroupRead

# R1 F-018: registry lookup 結果。exact=hit / none=未登録 / invalid=domain 正規化不能。
DomainTrustMatchType = Literal["exact", "none", "invalid"]


class ConflictCandidate(BaseModel):
    """R1 F-008: contradicting evidence を持つ claim (= 争点)。"""

    model_config = ConfigDict(from_attributes=True)

    claim_id: UUID
    contradicting_count: int
    supporting_count: int
    context_count: int
    conflict_group_id: UUID | None


class ClaimFreshness(BaseModel):
    """R1 F-006/F-009: claim の computed_freshness (read-only advisory)。"""

    claim_id: UUID
    computed_freshness: float | None
    newest_evidence_at: datetime | None


class EvidenceDomainTrust(BaseModel):
    """R1 F-018: evidence source の domain に対する trust registry 適用結果。"""

    evidence_source_id: UUID
    domain: str | None
    trust_tier: TrustTier | None
    match_type: DomainTrustMatchType


class ResearchAdvancedSummary(BaseModel):
    """research_task の advanced 集約 read model。"""

    research_task_id: UUID
    conflict_groups: list[ConflictGroupRead]
    conflict_candidates: list[ConflictCandidate]
    # relation_coverage = relation 付き evidence を持つ claim 比率 (0.0-1.0)。
    # 「争点なし」と「relation 未整備」を UI が区別するための signal。
    relation_coverage: float
    claim_freshness: list[ClaimFreshness]
    evidence_domain_trust: list[EvidenceDomainTrust]


class ConflictCandidateListResponse(BaseModel):
    items: list[ConflictCandidate]
    relation_coverage: float


__all__ = [
    "ClaimFreshness",
    "ConflictCandidate",
    "ConflictCandidateListResponse",
    "DomainTrustMatchType",
    "EvidenceDomainTrust",
    "ResearchAdvancedSummary",
]
